File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/InventoryQuery/Sorts/SortByStarRatingDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.InventoryQuery.Sorts;

/// <summary>
///     A decorator that sorts inventory items by star rating.
/// </summary>
public sealed class SortByStarRatingDecorator : InventoryQueryDecoratorBase
{
    private readonly SortDirection _direction;

    /// <summary>
    ///     Initializes a new instance of the <see cref="SortByStarRatingDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="direction">The sort direction.</param>
    public SortByStarRatingDecorator(IInventoryQuery inner, SortDirection direction = SortDirection.Ascending)
        : base(inner)
    {
        _direction = direction;
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
        => $"Sort: Star rating ({_direction.ToString().ToLowerInvariant()})";

    /// <inheritdoc />
    protected override IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items)
    {
        var sorted = _direction == SortDirection.Ascending
            ? items.OrderBy(i => i.StarRating.Rating)
            : items.OrderByDescending(i => i.StarRating.Rating);

        return sorted.ToList().AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/sorts/SortByStarRatingDecorator.java
package assignment2solution.domain.inventoryquery.sorts;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * A decorator that sorts inventory items by star rating.
 */
public final class SortByStarRatingDecorator extends InventoryQueryDecoratorBase {
    private final SortDirection direction;

    /**
     * Initializes a new instance of the {@link SortByStarRatingDecorator} class.
     */
    public SortByStarRatingDecorator(IInventoryQuery inner, SortDirection direction) {
        super(inner);
        this.direction = direction == null ? SortDirection.ASCENDING : direction;
    }

    public SortByStarRatingDecorator(IInventoryQuery inner) {
        this(inner, SortDirection.ASCENDING);
    }

    @Override
    protected String getAppliedDescription() {
        return "Sort: Star rating (" + direction.name().toLowerCase(Locale.ROOT) + ")";
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        var comparator = Comparator.comparingInt((InventoryItem i) -> i.getStarRating().getRating());
        if (direction == SortDirection.DESCENDING) {
            comparator = comparator.reversed();
        }
        return items.stream().sorted(comparator).collect(Collectors.toUnmodifiableList());
    }
}
